src/Processors/Pipe.cpp

```diff
@@ -250,6 +251,44 @@ static Pipes removeEmptyPipes(Pipes pipes)
     return res;
 }
 
+static Block getCommonHeader(const Pipes & pipes)
+{
+    Block res;
+
+    for (const auto & pipe : pipes)
+    {
+        if (const auto & header = pipe.getHeader())
+        {
+            res = header;
+            break;
+        }
+    }
+
+    for (const auto & pipe : pipes)
+    {
+        const auto & header = pipe.getHeader();
+        for (size_t i = 0; i < res.columns(); ++i)
+        {
+            /// We do not check that headers are compatible here. Will do it later.
+
+            if (i >= header.columns())
+                break;
+
+            auto & common = res.getByPosition(i).column;
+            const auto & cur = header.getByPosition(i).column;
+
+            /// Only remove const from common header if it is not const for current pipe.
+            if (cur && common && !isColumnConst(*cur))
```

Comment:
Can `common` be const?

Reply:
Yes, that's the point.
This code removes const from common header if in different pipes `const ` is different for the same columns.

Reply:
So is is possible, that column in common header is const, columns in all pipes are const, but const columns contain different values?

Reply:
For pipe this is not possible and separately checked.
For query plan, this is possible, and different constants are separately materialized.